### ISSUE TITLE
`build.sh` fails when run in a directory that contains spaces

### ISSUE THREAD

**Issue by MichalPetryka** [dotnet/runtime#73327]
Running in such directory results in:
```
<path with a space>/runtime/eng/build.sh: line 156: <path with a space up to the space>: No such file or directory
```
It looks like this line is the cause:
https://github.com/dotnet/runtime/blob/main/eng/build.sh#L156

**Comment by MichalPetryka** [comment]
It seems like fixing that line leads to more errors later on.

**Comment by ViktorHofer** [comment]
Thanks for brining this to our attention. Would you mind submitting a PR with the fix? And can you please post the failures that happen when this line is fixed?

**Comment by jkotas** [comment]
Related:
https://github.com/dotnet/runtime/issues/42397
https://github.com/dotnet/runtime/issues/13740

**Comment by MichalPetryka** [comment]
> Thanks for brining this to our attention. Would you mind submitting a PR with the fix? And can you please post the failures that happen when this line is fixed?

I've already moved my repo to a path without spaces and closed the terminal so I can't tell exactly but it was an error when invoking `msbuild`, so changing that one line for sure is not enough. I'll try fixing it again tomorrow.

**Comment by MichalPetryka** [comment]
> Thanks for brining this to our attention. Would you mind submitting a PR with the fix? And can you please post the failures that happen when this line is fixed?

Looked into this again, the following error appears after fixing the first one, I don't really know how to fix it though:
```sh
MSBUILD : error MSB1008: Only one project can be specified.
 Full command line: „/media/ewa/Samsung USB/runtime/.dotnet/sdk/7.0.100-preview.7.22377.5/MSBuild.dll -maxcpucount -verbosity:m /m /nologo /clp:Summary /v:minimal /nr:true /warnaserror /p:TreatWarningsAsErrors=true /p:ContinuousIntegrationBuild=false /home/ewa/.nuget/packages/microsoft.dotnet.arcade.sdk/7.0.0-beta.22416.1/tools/Build.proj /p:Configuration=Release /p:RepoRoot=/media/ewa/Samsung USB/runtime/ /p:Restore=true /p:Build=true /p:ArcadeBuildFromSource=false /p:Rebuild=false /p:Test=false /p:Pack=false /p:IntegrationTest=false /p:PerformanceTest=false /p:Sign=false /p:Publish=false /p:Subset=clr+libs /p:TargetArchitecture=x64 /p:BuildArchitecture=x64 /p:CMakeArgs="" -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,/media/ewa/Samsung USB/runtime/.dotnet/sdk/7.0.100-preview.7.22377.5/dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,/media/ewa/Samsung USB/runtime/.dotnet/sdk/7.0.100-preview.7.22377.5/dotnet.dll“
 Switches appended by response files: 
Switch: USB/runtime/

For switch syntax, type "MSBuild -help"
```

**Comment by jkotas** [comment]
Try to wrap https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/eng/common/build.ps1#L113 in quotes, like ` "/p:RepoRoot=$RepoRoot"`

**Comment by MichalPetryka** [comment]
> Try to wrap
> 
> https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/eng/common/build.ps1#L113
> 
> in quotes, like ` "/p:RepoRoot=$RepoRoot"`

Already tried, the same error.

**Comment by ivdiazsa** [comment]
The MSBuild error is just part of the problem. Bash also doesn't like paths with spaces when sourcing other scripts.

For example, `build.sh --help` also fails, albeit with a different error. It treats the path as up to the first space, and consequently, there is no directory under that name. This one is fixed by enclosing `"$scriptroot"` with quotes. Most of our scripts already did that but there were a few unquoted instances.

Currently looking into the MSBuild failure as Jan's suggestion unfortunately didn't work as expected.

**Comment by ivdiazsa** [comment]
After some more investigation, and reading through the issues pointed by @jkotas here https://github.com/dotnet/runtime/issues/73327#issuecomment-1204400327, I can conclude this issue goes far deeper than meets the eye. It might actually be a bug with the MSBuild parser rather than our runtime, since it just ignores my quoted paths and splits in spaces. Additionally, I got two different repros in MacOS/Linux and Windows:

**Linux/MacOS**

```bash
__DistroRid: osx-arm64
MSBUILD : error MSB1008: Only one project can be specified.
    Full command line: '/usr/local/share/dotnet/sdk/7.0.100-preview.7.22377.5/MSBuild.dll -maxcpucount -verbosity:m /m /nologo /clp:Summary /v:minimal /nr:true /warnaserror /p:TreatWarningsAsErrors=true /p:ContinuousIntegrationBuild=false /Path/To/.nuget/packages/microsoft.dotnet.arcade.sdk/7.0.0-beta.22419.1/tools/Build.proj /p:Configuration=Release /p:RepoRoot="/Path/To/Runtime With Spaces/" /p:Restore=true /p:Build=true /p:ArcadeBuildFromSource=false /p:Rebuild=false /p:Test=false /p:Pack=false /p:IntegrationTest=false /p:PerformanceTest=false /p:Sign=false /p:Publish=false /p:Subset=clr /p:TargetArchitecture=arm64 /p:BuildArchitecture=arm64 /p:CMakeArgs="" -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,/usr/local/share/dotnet/sdk/7.0.100-preview.7.22377.5/dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,/usr/local/share/dotnet/sdk/7.0.100-preview.7.22377.5/dotnet.dll'
  Switches appended by response files:
Switch: With

For switch syntax, type "MSBuild -help"
Build failed with exit code 1. Check errors above.
```

Check out the `/p:RepoRoot` flag. Also, out of curiosity, I passed the path with just a starting quote, but no closing one, and it failed with the exact same error. Which means it ignores them. Single quotes and no quotes yielded the same result.

**Windows**

```cmd
C:\Path\To\Runtime With Spaces\src\coreclr\runtime.proj(61,5): error MSB3073: The command ""C:\Path\To\Runtime With Spaces\src\coreclr\build-runtime.cmd" -x64 -release -os windows -pgodatapath "C:\Path\To\.nuget\packages\optimization.windows_nt-x64.pgo.coreclr\1.0.0-prerelease.22415.6"" exited with code 1.

Build FAILED.

C:\Path\To\Runtime With Spaces\src\coreclr\runtime.proj(61,5): error MSB3073: The command ""C:\Path\To\Runtime With Spaces\src\coreclr\build-runtime.cmd" -x64 -release -os windows -pgodatapath "C:\Path\To\.nuget\packages\optimization.windows_nt-x64.pgo.coreclr\1.0.0-prerelease.22415.6"" exited with code 1.
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:13.20
Build failed with exit code 1. Check errors above.
```

I found the offending line in `build-runtime.cmd` and fixed it (there was a `call` statement without quoting the script path name), and then it failed later on somewhere else. My hypothesis is the problem will keep propagating to other places (our scripts' problem) until it eventually lands on the MSBuild problem like it happens in MacOS and Linux.

**Comment by trylek** [comment]
This is a bottomless Pandora's box. Manish also spent some time on this but to my knowledge no-one has yet made this work end to end. I think that in general it's a worthwhile effort so if you have some fixes that let the build proceed further, just get them merged in; once we're sure we're no longer able to make progress due to internal issues in components like MSBuild, we can follow up with the partner teams.

**Comment by danmoseley** [comment]
Building with spaces seems to fail about once a year. Which tells me that the developers working on this day to day just avoid it and it will keep breaking periodically unless we make CI work this way.

We could just document that we don't support it...

**Comment by trylek** [comment]
Sure, we all strive to avoid it, but some people are less lucky - in many institutions, universities, companies, IT may force a policy that requires people to use root folders with names beyond their control (e.g. based on their names that may contain a plethora of accent signs and spaces). While it's a sad reality that today this doesn't work, I think our north star should be ultimately making this fully functional, at the very least as a nod to our inclusiveness principle.

**Comment by trylek** [comment]
And, after all, the default VS installation folder contains at least four blanks (<code>C:\Program Files (x86)\Microsoft Visual Studio</code>).

**Comment by MichalPetryka** [comment]
> Building with spaces seems to fail about once a year. Which tells me that the developers working on this day to day just avoid it and it will keep breaking periodically unless we make CI work this way.
> 
> We could just document that we don't support it...

There could be some weekly CI job that'd build and test from a directory that contains spaces and non-ASCII characters, this'd keep things working in check.

**Comment by trylek** [comment]
(In practice, even this is non-trivial as AzDO has its own rules regarding directory naming, I believe we looked into that in the past and there was no easy way to name the AzDO folder for a particular run "šílený raťafák" so that would be one more thing to follow up on with the AzDO team.)

**Comment by danmoseley** [comment]
You're right, now I recall in the past at least one of the folks hitting this had no other option, because their user name was chosen to have a space and they could not access any other folder.

**Comment by akoeplinger** [comment]
One workaround for those users could be to build in a VM/Docker container.

**Comment by danmoseley** [comment]
It's possible the machine is locked down such that they can't do that kind of thing?

**Comment by ivdiazsa** [comment]
> This is a bottomless Pandora's box. Manish also spent some time on this but to my knowledge no-one has yet made this work end to end. I think that in general it's a worthwhile effort so if you have some fixes that let the build proceed further, just get them merged in; once we're sure we're no longer able to make progress due to internal issues in components like MSBuild, we can follow up with the partner teams.

Alright then it's settled! I'll get a PR with fixes on our part, i.e. until the failure falls onto MSBuild.

**Comment by ivdiazsa** [comment]
> It's possible the machine is locked down such that they can't do that kind of thing?

This is a thing in some companies so unfortunately that workaround wouldn't work for them. It's not a huge amount of places, but there are enough locked down that warrants looking into this.

**Comment by ApparentlyPlus** [comment]
This has been quiet for a while, but I hit this exact issue today on my box and checked to see if it was documented, which brought me here. It seemed interesting, so I dug into it to see how far the rabbit hole gets.

Everything you see below was reproduced locally on linux-x64 (Fedora 44), in a worktree at a path *containing a space*. Take all of this with a grain of salt, since I didn't test any fixes on windows or macOS at all. Also as a side note, the line referenced in the discussion above is outdated, now [`eng/build.sh:169`](https://github.com/dotnet/runtime/blob/838818c950f2448767d5feae460036e3f88e093d/eng/build.sh#L169) rather than 156.

In the time I spent digging, I was able to find 5 failure points downstream, and locally fix the first four, which gets `./build.sh -subset libs.native -c Release` to exit 0 at a path with a space. Here's the gist of it:

| # | Location | Cause |
| --- | --- | --- |
| 1 | [`eng/build.sh:169`](https://github.com/dotnet/runtime/blob/838818c950f2448767d5feae460036e3f88e093d/eng/build.sh#L169) | unquoted `source $scriptroot/...` |
| 2 | [`eng/common/build.sh:277`](https://github.com/dotnet/runtime/blob/838818c950f2448767d5feae460036e3f88e093d/eng/common/build.sh#L277) | `MSBuild $_InitializeToolset` word-splits, giving `MSB1008: Only one project can be specified` |
| 3 | [`eng/native/functions.cmake:265-276`](https://github.com/dotnet/runtime/blob/838818c950f2448767d5feae460036e3f88e093d/eng/native/functions.cmake#L265-L276) | unquoted `${exports_filename}` in `set_exports_linker_option` |
| 4 | [`System.IO.Compression.Native/CMakeLists.txt:95`](https://github.com/dotnet/runtime/blob/838818c950f2448767d5feae460036e3f88e093d/src/native/libs/System.IO.Compression.Native/CMakeLists.txt#L95) | `LINK_FLAGS` cannot carry a path with spaces |
| 5 | [`src/coreclr/build-runtime.sh:176`](https://github.com/dotnet/runtime/blob/838818c950f2448767d5feae460036e3f88e093d/src/coreclr/build-runtime.sh#L176) | path is quoted, but `eval` re-parses the string and splits on the space |

In light of this, I think I now get why this hasn't moved forward in a while:

1. Part of the fix is not in this repo, specifically, Item 2. It lives in `eng/common/`, which afaik is mirrored from `dotnet/arcade` and is explicitly off limits here. So even the minimal path to a working `libs.native` needs an arcade PR and a version flow before the runtime side can be completed. I did not see this mentioned anywhere in the thread.

2. Item 4 is actually not a quoting problem. `LINK_FLAGS` is a raw string that CMake passes to the linker as is, so quoting the variable in CMake source changes nothing. I confirmed that and got the identical `clang: no such file or directory` error. What does work is switching to `target_link_options`, which CMake escapes properly:

    ```diff
    -set_property(TARGET X APPEND_STRING PROPERTY LINK_FLAGS ${EXPORTS_LINKER_OPTION})
    +target_link_options(X PRIVATE "${EXPORTS_LINKER_OPTION}")
    ```

I'm mentioning this specifically for anyone who tried solving this before, since quoting looks like the obvious fix and produces no visible change as far as I could tell. On top of that, I confirmed 10 sites using this exact exports pattern and 23 `APPEND_STRING PROPERTY LINK_FLAGS` uses in total.

Is it a pandora's box like @trylek said? Kinda, but the scope isn't insane. The fix seems to be a finite set of mechanical patterns, spread across two repos and most native components. The pattern was unquoted `$var` in shell, unquoted `${VAR}` in CMake, and the `LINK_FLAGS` to `target_link_options` migration.

@ivdiazsa is still assigned from 2022. I wonder whether anyone is actively on this? Given @trylek's earlier point about landing partial fixes that let the build proceed further, I would be happy to help out incrementally, starting with the `libs.native` fix above, if that is still the preferred approach and nobody else has it in flight.

What I am most worried about is regressions and testing, especially since this could be breaking. @danmoseley already said that nothing in CI covers this, so it will keep regressing regardless of how far the fixes get.

**Comment by akoeplinger** [comment]
@ApparentlyPlus we can make incremental progress :) the first two are easy ones and can be just fixed, one in dotnet/runtime and one in dotnet/arcade. Feel free to send a PR.

**Comment by ApparentlyPlus** [comment]
> [@ApparentlyPlus](https://github.com/ApparentlyPlus) we can make incremental progress :) the first two are easy ones and can be just fixed, one in dotnet/runtime and one in dotnet/arcade. Feel free to send a PR.

Awesome. I'll take this up asap. I'll draft a PR either tonight or tomorrow morning.

**Comment by ApparentlyPlus** [comment]
@akoeplinger Apologies for being a tad bit late, I needed to run a lot of things to understand the full shape of this. Let me also apologize in advance for how long this message is, but the issue warrants it, and it's a pretty fascinating one for me :D. 

So, both PRs for the first tier are up:

- dotnet/arcade: [dotnet/arcade#17509](https://github.com/dotnet/arcade/pull/17509)
- dotnet/runtime: [#133424](https://github.com/dotnet/runtime/pull/133424)

Between them they get the build past the shell layer at a path containing a space, on Linux *and* macOS (and presumably all unix targets). They do not fully solve this. Windows is a different discussion, more on that later. 

### TL;DR

- The two PRs fix **five** separate space path failures in the shell layer, three in arcade and two in runtime. Verified on Linux and macOS.
- They **do not close this**. Once they land the build gets to the CMake tier and stops on three more things.
- My earlier five point list was wrong in places. One item needed no fix at all, one was non fatal, one count was off, and it missed three failures entirely because they only show up behind flags like `--clean` and `--binaryLog`.
- The reason this recurs is structural: [`build-commons.sh`](https://github.com/dotnet/runtime/blob/040defaa3b808c418df16029e98c5af3b5273583/eng/native/build-commons.sh#L209) builds the whole cmake invocation as one string and `eval`s it, so every caller has to remember to quote its own arguments. Miss one and it breaks again.
- What I would like to propose: the remaining CMake fixes and the `eval` removal land **together**, validated by existing CI staying green plus hand testing on every Unix platform, which I will post here before opening anything.
- Windows is the same bug class but a separate discussion, tracked in [#42397](https://github.com/dotnet/runtime/issues/42397) and [#13740](https://github.com/dotnet/runtime/issues/13740).

### What these two fix

Every failure in this issue is the same mistake wearing a different coat. A path gets passed around as a bare `$variable` instead of `"$variable"`, the shell splits it on the space, and `/home/me/space test/runtime` arrives at whatever consumed it as two separate arguments. That is the whole issue, in a nutshell. Each occurrence needs the same one line fix, and the only reason this looks bottomless is that fixing one just gets you as far as the next one. Well, until you fix them all.

**The arcade PR** ([#17509](https://github.com/dotnet/arcade/pull/17509)) clears three of them:

- Every build dies at [`MSBuild $_InitializeToolset`](https://github.com/dotnet/arcade/blob/20d75e9daf1ed4f6509ca04436842a1b633f02eb/eng/common/build.sh#L277) with MSB1008.
- `--binaryLog` dies the same way at [`$bl`](https://github.com/dotnet/arcade/blob/20d75e9daf1ed4f6509ca04436842a1b633f02eb/eng/common/build.sh#L269), *even after* the above is fixed.
- `--clean` reports success and deletes nothing at [`rm -rf $artifacts_dir`](https://github.com/dotnet/arcade/blob/20d75e9daf1ed4f6509ca04436842a1b633f02eb/eng/common/build.sh#L302).

**The runtime PR** ([#133424](https://github.com/dotnet/runtime/pull/133424)) clears two:

- Every build dies at [`eng/build.sh:169`](https://github.com/dotnet/runtime/blob/040defaa3b808c418df16029e98c5af3b5273583/eng/build.sh#L169), which is the failure originally reported here.
- `-bootstrap` dies at [`eng/build.sh:647`](https://github.com/dotnet/runtime/blob/040defaa3b808c418df16029e98c5af3b5273583/eng/build.sh#L647).

I hit them in the same order as above, and together the two take you from "nothing runs at all" to the CMake tier. Note that runtime only picks up the arcade fix **once a fresh arcade mirror flows in** (after the PR lands there), so the two do not land at the same moment.

### Things I got wrong

Now that I picked up the issue and did a proper deep dive, I uncovered that parts of what I said earlier were wrong. I listed five failure points. The list was right in shape but wrong in three places.

Item 3, the unquoted `${exports_filename}` in [`set_exports_linker_option`](https://github.com/dotnet/runtime/blob/040defaa3b808c418df16029e98c5af3b5273583/eng/native/functions.cmake#L265-L276), does not need fixing at all. I listed it as one of the four fixes I had tried, and it is a no op. Item 3 and 4 were related and technically both true, but they only required one fix. CMake splits unquoted argument expansions on semicolons only, never on whitespace, so adding quotes there changes absolutely nothing. I checked it both directions. Fixing only item 4 makes the build pass, and fixing only item 3 gives the *identical* `clang: no such file or directory`. My later remark that quoting it produces no visible change was right. Calling it a fix however, was just plain wrong.

Item 5, the [`eval` in `build-runtime.sh`](https://github.com/dotnet/runtime/blob/040defaa3b808c418df16029e98c5af3b5273583/src/coreclr/build-runtime.sh#L176), is real and it does fire at exactly that line, but it is apparently **non fatal**. The script prints the error and carries on regardless, so it was never the thing blocking anyone. Same bug is duplicated at [`eng/native/build-commons.sh:52`](https://github.com/dotnet/runtime/blob/040defaa3b808c418df16029e98c5af3b5273583/eng/native/build-commons.sh#L52), btw. This might be a nice-to-have change that we could fold in, if you wish. Related, but not an issue.

And my count of "10 sites using this exact exports pattern" was again wrong. It is 9 for the exact pattern, or 13 if you include the renamed variants (`JIT_`, `SPMI_COLLECTOR_`, and the two numbered ones in palsuite). The 23 figure for `APPEND_STRING PROPERTY LINK_FLAGS` overall was correct.

### Why my list was the wrong shape

Beyond the three mistakes above, my list had a structural problem: everything on it was something that breaks a plain `build.sh` invocation. But three of the five shell failures only surface behind a flag, `--binaryLog`, `--clean` and `-bootstrap`, so running the build until it stops erroring will never find them. All three are fixed by the two PRs above, but only because I went looking for them specifically. That is worth knowing for scoping this, and probably explains why previous passes at it also came up short.

`--clean` was the one I liked least. It returned exit 0 and cheerfully printed "Artifacts directory deleted" while deleting nothing at all. Everything else here failed loudly, that one just lied to you. An unquoted `rm -rf` over a split path could also take out something you did want, if one of the prefixes happens to exist. I genuinely hope that has never bitten anyone.

### Why this keeps coming back

This is the gist of everything. [`eng/native/build-commons.sh`](https://github.com/dotnet/runtime/blob/040defaa3b808c418df16029e98c5af3b5273583/eng/native/build-commons.sh#L209) assembles the entire cmake invocation into one string and runs `eval` on it. The fixed paths carry embedded escaped quotes so they survive the re-parse, but `$cmakeArgs` gets appended raw, which means every producer has to remember to embed its own quoting. That is exactly why [`runtime.proj:56`](https://github.com/dotnet/runtime/blob/040defaa3b808c418df16029e98c5af3b5273583/src/coreclr/runtime.proj#L56) carries a single-quote workaround and [line 57](https://github.com/dotnet/runtime/blob/040defaa3b808c418df16029e98c5af3b5273583/src/coreclr/runtime.proj#L57), added later by someone who had no reason to know, does not.

So the design makes the broken thing the default and the correct thing an act of memory. Fixing that properly means moving `cmakeArgs` to an array and dropping the `eval`, which touches every native build entry point including mobile and wasm. I am proposing exactly that, just not in this pass. It is far too big to smuggle into a quoting PR, and it needs something watching it before it goes in. See below.

### What I actually verified

Linux x64 on Fedora 44 and macOS 15.7.3 on Apple Silicon, both in worktrees at a path containing a space.

Every one of the five shell fixes is independently necessary. I ran all 32 combinations of them against a plain build, `--binaryLog` and `--clean`, 96 runs in total, and each component gates exactly one failure with no interaction between them.

For regression I built `clr.native+libs.native` clean at a space free path with and without the changes, in Debug and Release. The binary inventory and every exported symbol set come out identical, 118 binaries and 1318 symbols across 24 libraries, so none of this changes anything on a normal path. `host.native` and `mono.runtime` were checked too.

With the full set applied, including the native fixes that are *not* in these PRs, `clr.native`, `libs.native`, `host.native` and `mono.runtime` all build clean at a space path on linux-x64, 2877 ninja targets and no failures. On macOS the build reaches the identical exports failure through the `LD_OSX` branch instead of the Linux one, which at least confirms the same defect exists on both.

**One caveat on all of that.** There is no test suite for any of this. Every build path in this repo and in arcade exercises space free directories, so a regression produces no signal whatsoever, and the only detection mechanism is a contributor tripping over it on their own machine and opening an issue. That is precisely the history of this bug. Everything above is a manual A/B on two boxes, which is the best evidence available today, and it is why I want the remaining work validated by hand across every platform before it goes anywhere near a PR.

### What happens next

Once both PRs land, the build gets as far as the CMake tier, and stops there on three things: an unquoted cmake argument at [`src/coreclr/runtime.proj:57`](https://github.com/dotnet/runtime/blob/040defaa3b808c418df16029e98c5af3b5273583/src/coreclr/runtime.proj#L57) that aborts configure outright, the [`LINK_FLAGS`](https://github.com/dotnet/runtime/blob/040defaa3b808c418df16029e98c5af3b5273583/src/native/libs/System.IO.Compression.Native/CMakeLists.txt#L95) exports sites, and two separate quoting problems in the CMake [`COMMAND`](https://github.com/dotnet/runtime/blob/040defaa3b808c418df16029e98c5af3b5273583/src/coreclr/nativeaot/Runtime/Full/CMakeLists.txt#L165) that generates `AsmOffsets`: a shell redirect `>` sitting inside `add_custom_command`, and hand written quotes inside an argument, `-I"${NATIVEAOT_RUNTIME_DIR}"`. Both end up double escaped, because CMake escapes the space itself while the quotes you wrote survive as literal characters, so the compiler is handed `space\ test` with a real backslash in it and cannot find the file.

Those 14 `CMakeLists.txt` files need fixing on their own merits: `LINK_FLAGS` and the `AsmOffsets` commands fail after the arguments have already reached CMake intact, so no amount of tidying the shell layer touches them. But fixing only those puts us back where we started, a pile of patches with nothing stopping the next one from landing unquoted. So the shape I would like to propose is:

1. **The CMake fixes and the array change should land together.** The array change replaces the flat `$cmakeArgs` string and the `eval` in `build-commons.sh` so that quoting becomes structural rather than something each caller has to remember, which is what stops the *next* one appearing. I want to be upfront that this is the breaking one. It touches every native build entry point including mobile and wasm, and it is the change that could plausibly regress something.

2. **Validation is existing CI plus local testing, not a new job.** I originally wanted a periodic job building from a space path to go in first, but I have been talked out of it and I think correctly. It would be expensive, it would be red from the day it lands until the work is finished, it would therefore have to be non-blocking, and once the fix is in it earns its keep only rarely. So instead, the existing CI legs have to stay green, which is what actually proves no regression for the overwhelming majority who have no spaces in their paths, and I do the space path validation by hand across every platform and report it here before opening anything.

3. **Then this can close.** Breaking change in, existing CI green, and local end to end results posted for each platform. That is a real ending.

A note on Windows, since these two PRs are Unix only and that is not an oversight. PowerShell does not word split on variable expansion, and there is no `Invoke-Expression` anywhere in arcade's `eng/common/*.ps1`, so `build.ps1` simply does not have this bug class. There is nothing to mirror. The real Windows exposure is elsewhere: the `.cmd` batch files, where `%VAR%` genuinely does split, and the CMake `COMMAND` blocks that shell out to `lib.exe` and powershell.

That side of it is tracked separately in [#42397](https://github.com/dotnet/runtime/issues/42397), which is the same bug class in `build.cmd`, and [#13740](https://github.com/dotnet/runtime/issues/13740) for the `ilasm` variant. Related, but not this issue, so I would rather not drag them in here beyond noting that the CMake tier above is shared ground and whatever lands for it has to work on all three platforms.

Worth naming one thing from #42397 though, because it is the whole argument in miniature. It was already fixed once, in [#42671](https://github.com/dotnet/runtime/pull/42671), and @mangod9 notes on that thread that there is a good chance things have since regressed. A fix landed, nothing watched it, and five years later nobody can say whether it still holds. That is exactly the outcome I want to avoid here, and it is why I would rather post per platform results than merge on the strength of a green CI run that never touches a path with a space in it.

Which is the last thing. Before I open anything further I will run the full set end to end on both Unix platforms and both architectures, so linux-x64, linux-arm64, macos-arm64 and macos-x64, and post the results here first. Everything above is linux-x64 and macos-arm64 only so far. I will cover Windows too, since the CMake tier is shared ground, but those results belong in #42397 rather than here.

On that note, since this is unassigned and I have the next tier mapped out, could I be assigned to it? Happy to keep it moving, and equally happy to hand it back if someone on the team would rather own it.

**Comment by ApparentlyPlus** [comment]
Thanks for the assignment teo. I'm waiting on review in the 2 open PRs and someone to give me the green light on phase 2, or vehemently disagree, which I welcome all the same 👍🏻

**Comment by ApparentlyPlus** [comment]
Jesus christ. I was completely wrong, and @trylek was right in 2022. This is a pandora's box. Forget everything I said above. This goes way deeper than I gave it credit for. I'm waiting to coordinate on the open PRs first before I dive deeper.

**Comment by ApparentlyPlus** [comment]
Howdy again, and thanks for merging both PRs @akoeplinger! Shorter comment this time, I promise. I deleted my prior ones because they were absolute and misleading, so here's a revised, hopefully better assessment.

Since iterating is back on the table, I mapped the rest of it by fixing faults one at a time, until I managed to make it compile locally. I got a full `./build.sh -c Release` to complete from scratch at a path with a space on **linux-x64 and macos-arm64**. This is good news, but I am not gonna claim what I have is the full scope. Every time I did so far, more things surfaced.

Turns out, this is almost purely a quoting problem. Every failure downstream is the same mistake repeated: a path handed off as a bare `$var` / `$(Prop)` / `%(Meta)` into something that word splits, and each one eluded me until the one before it was fixed. That's why the file count skyrocketed.

A full fix breaks down in 3 steps:

### 1. We need another arcade PR, this time for the installers package

Still a quoting problem, and macOS needs this. When the build packages the product on macOS it shells out to `pkgbuild` and `productbuild`. Those are macOS system tools, but the targets driving them aren't in this repo, they ship inside the `Microsoft.DotNet.Build.Tasks.Installers` NuGet package, which comes from arcade. Both assemble a flat command string and hand it to `Exec`, so every path in it needs quoting:

- [`installer.build.targets:449-455`](https://github.com/dotnet/arcade/blob/08fb2ee2de703bd5b3eb6653aa8814d9de867112/src/Microsoft.DotNet.Build.Tasks.Installers/build/installer.build.targets#L449-L455), `pkgbuild`
- [`bundle.targets:177-183`](https://github.com/dotnet/arcade/blob/08fb2ee2de703bd5b3eb6653aa8814d9de867112/src/Microsoft.DotNet.Build.Tasks.Installers/build/bundle.targets#L177-L183), `productbuild`

>[!Important]
> This is not `eng/common`, so there is no mirror. It has to merge in arcade, get a package release, and then that version has to flow back into runtime before macOS CI can go green. That's gonna take a while (I'd love it if someone shed some light on how long), and nothing else waits on it, so it should go up first while the rest moves.

### 2. The `eval` in `build-commons.sh` should change

[`build-commons.sh:209`](https://github.com/dotnet/runtime/blob/0a719e7753978ec8ad21c7b7fd41bddcc723d2ce/eng/native/build-commons.sh#L209) plus the four callers. It's different in kind from the rest, and it has to happen before the other quoting fixes.

The script builds the whole `gen-buildsys.sh` invocation as a *string* and runs it through `eval`, so it gets parsed twice and `$cmakeArgs` splits on that second pass. You can't fix it from the calling side either, since a caller's quotes get eaten by the first parse. The fix is to drop the string, `cmakeArgs` becomes a bash array and `gen-buildsys.sh` is invoked directly with `"${cmakeArgs[@]}"`.

Btw, this is the only real behaviour change in the set, since some values leaned on that second parse. The iOS [`armv7;armv7s`](https://github.com/dotnet/runtime/blob/0a719e7753978ec8ad21c7b7fd41bddcc723d2ce/src/native/libs/build-native.sh#L94) one splits on the `;` and would have taken out the armv7 leg if I hadn't re-quoted it. It's risky, but anything I miss, the current CI will catch.

I built both halves separately to check the order. Quoting without this dies at CMake configure before a single object compiles, on [a line that is already quoted correctly](https://github.com/dotnet/runtime/blob/0a719e7753978ec8ad21c7b7fd41bddcc723d2ce/src/coreclr/clrdatadescriptors.cmake#L35-L37), because the `eval` ate the `-D` before CMake saw it. The `eval` alone gets 1100 of 2877 objects in and dies at link. So, quoting first would look like it did nothing.

### 3. Quotes everywhere, a full sweep of the same thing

This is 21 files, but they're all the same fix:

- CMake `LINK_FLAGS` / `COMMAND`, 14 files (eg. [one of them](https://github.com/dotnet/runtime/blob/0a719e7753978ec8ad21c7b7fd41bddcc723d2ce/src/coreclr/dlls/mscordac/CMakeLists.txt#L129))
- MSBuild `Exec`, 6 files: [`crossgen-corelib.proj`](https://github.com/dotnet/runtime/blob/0a719e7753978ec8ad21c7b7fd41bddcc723d2ce/src/coreclr/crossgen-corelib.proj#L114-L117) twice (mibc merge, then crossgen2), [`externals.csproj:81`](https://github.com/dotnet/runtime/blob/0a719e7753978ec8ad21c7b7fd41bddcc723d2ce/src/libraries/externals.csproj#L81), [`Crossgen2.sfxproj:17`](https://github.com/dotnet/runtime/blob/0a719e7753978ec8ad21c7b7fd41bddcc723d2ce/src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj#L17), [`dotnet-host.proj:174`](https://github.com/dotnet/runtime/blob/0a719e7753978ec8ad21c7b7fd41bddcc723d2ce/src/installer/pkg/sfx/installers/dotnet-host.proj#L174), `mono.proj` and `native-library.targets`
- [`dotnet.sh`](https://github.com/dotnet/runtime/blob/0a719e7753978ec8ad21c7b7fd41bddcc723d2ce/dotnet.sh#L18-L24), three bugs in 24 lines

Oh, and coincidentally, [`sdk-task.sh:37-41`](https://github.com/dotnet/arcade/blob/08fb2ee2de703bd5b3eb6653aa8814d9de867112/eng/common/sdk-task.sh#L37-L41) is the same `$binaryLogArg` bug dotnet/arcade#17509 fixed in `build.sh` this morning, sitting one file over. I didn't see it before, otherwise I'd have fixed that too. Again, quotes.

So, the arcade PR should go up next, along with the `eval` fix here. If you green light it, that's another 2 relatively small PRs, then the quoting sweep and some patience until the arcade package flows into runtime. Bear in mind that green CI only confirms no regressions, not the fix itself. The fix can only be tested locally, and it will break again unless we add CI after everything.

Waiting on the okay and I'm on it.

**Comment by akoeplinger** [comment]
@ApparentlyPlus 

1. arcade changes look reasonable to me, feel free to open a PR. once the change is merged there it will flow to dotnet/dotnet and then to dotnet/runtime, you can watch for codeflow PRs like https://github.com/dotnet/runtime/pull/133655
2. sounds reasonable. About the armv7/armv7s, that's actually dead code since we no longer build for ios/iossimulator for those platforms or i386 anymore. Logic like that can be cleaned up separately.
3. sure

**Comment by ApparentlyPlus** [comment]
@akoeplinger First 2 are up, one in arcade and one in runtime. I'll do the third one once these two are green and land. Thanks for the quick replies!
